### DISCUSSION TITLE
Removing -type-in-type

### DISCUSSION THREAD

**Discussion by ReimannJ** [UniMath/UniMath#1776] [General]
For a few weeks now, @benediktahrens , @m-lindgren , and I have been discussing different options on how to remove -type-in-type from UniMath. We have tried out different combinations of resizing rules (see e.g. [Voevodsky's slides](https://www.math.ias.edu/~vladimir/Site3/Univalent_Foundations_files/2011_Bergen.pdf) or [our preliminary table of potential combinations](https://github.com/UniMath/UniMath/wiki/Resizing-rules-for-removing-type%E2%80%90in%E2%80%90type)) but have not yet settled on a good combination of rules, or even on what "good" means in this context, so we wanted to open the discussion up to the community. 

- Does anyone already have any experience working with different resizing rules?

- Another option would be to use resizing axioms instead of resizing rules, similar to the [HoTT library](https://github.com/HoTT/Coq-HoTT/blob/master/theories/PropResizing/PropResizing.v). We have not tried this yet, but perhaps someone has experiences with/opinions on using resizing axioms?

- Any other ideas/concerns/things to keep in mind? 🙂

**Comment by nmvdw** [comment]
I have several questions:
- There are several resizing rules on the discussion page. What is their meta-theoretic status? Are they known to be consistent, and which rules hold in the simplicial set model?
- What does `Set` refer to on the page https://github.com/UniMath/UniMath/wiki/Resizing-rules-for-removing-type%E2%80%90in%E2%80%90type? 
- How difficult is it to implement each of the resizing rules?

The resizing axiom in the HoTT library is strong enough for constructing truncations (https://github.com/HoTT/Coq-HoTT/blob/6c2c3f6db2b14689dbb5338e7fc03230b8518c81/theories/PropResizing/ImpredicativeTruncation.v), and one can also construct quotients and a subobject classifier for the category of sets (see "Sets in homotopy type theory" by Rijke and Spitters).

The main points for me would be the following:
- The resizing rules that we take, should hold in the simplicial model
- I prefer a simpler implementation over a more complicated one, even if that makes propositional resizing more complicated to use. I think that resizing is mainly used for several basic constructions (such as truncations, quotients, and so on).

**Comment by DanGrayson** [comment]
Vladimir had no proof that the resizing rules were consistent.

I started working on this long ago (see my branch "file-by-file-type-in-type-universe-monomorphic"), but ran into difficulties.

**Comment by nmvdw** [comment]
I prefer that the usage of resizing becomes explicit. Not too long ago, I formalized a proof that if a DCPO is continuous (up to truncation), then one can construct a continuity structure for that DCPO. I am currently supervising a student who formalized that in Type Topology (https://github.com/martinescardo/TypeTopology/pull/206), and for that proof, resizing was necessary (which I was not aware of). By leaving the usage of resizing implicit, one is not aware whether resizing is used in a proof or not. If resizing is explicit (via an axiom expressed using equivalences), the prover is more aware of what axioms are necessary for a proof, which I think is a significant advantage.
